src/deps-index.ts

```diff
@@ -21,79 +20,86 @@ export interface DepsIndex {
+ */
+function getHtmlDependencies(document: Document): DependencyMapEntry {
+  const deps = new Set<UrlString>();
+  const eagerDeps = new Set<UrlString>();
+  const lazyImports = new Set<UrlString>();
+  _getHtmlDependencies(document, true, deps, eagerDeps, lazyImports);
+  return {deps, eagerDeps, lazyImports};
+}
 
-        switch (htmlImport.type) {
-          case 'html-import':
-            eagerImports.add(htmlImport.document.url);
-            break;
-          case 'lazy-html-import':
-            lazyImports.add(htmlImport.document.url);
-            break;
-        }
-      }
-      return {url: url, eager: eagerImports, lazy: lazyImports};
+function _getHtmlDependencies(
+    document: Document,
+    viaEager: boolean,
+    visited: Set<UrlString>,
+    visitedEager: Set<UrlString>,
+    lazyImports: Set<UrlString>) {
+  const htmlImports = document.getFeatures(
+      {kind: 'html-import', imported: false, externalPackages: true});
+  for (const htmlImport of htmlImports) {
+    const importUrl = htmlImport.document.url;
+    if (htmlImport.lazy) {
+      lazyImports.add(importUrl);
+    }
+    if (visitedEager.has(importUrl)) {
+      continue;
+    }
+    const isEager = viaEager && !htmlImport.lazy;
+    if (isEager) {
+      visitedEager.add(importUrl);
+      // In this case we've visited a node eagerly for the first time,
+      // so recurse
+    } else if (visited.has(importUrl)) {
+      // In this case we're seeing a node lazily again, so don't recurse
+      continue;
     }
+    visited.add(importUrl);
+    _getHtmlDependencies(
+        htmlImport.document, isEager, visited, visitedEager, lazyImports);
+  }
+}
 
+/**
+ * Analyzes all entrypoints and determines each of their transitive
+ * dependencies.
+ * @param entrypoints Urls of entrypoints to analyze.
+ * @param analyzer
+ * @return a dependency index of every entrypoint, including entrypoints that
+ *     were discovered as lazy entrypoints in the graph.
+ */
 export async function buildDepsIndex(
     entrypoints: UrlString[], analyzer: Analyzer): Promise<DepsIndex> {
-  const entrypointToDependencies: Map<UrlString, Set<UrlString>> = new Map();
-  const dependenciesToEntrypoints: Map<UrlString, Set<UrlString>> = new Map();
-  const queue = Array.from(entrypoints);
-  const visitedEntrypoints = new Set<UrlString>();
-  while (queue.length > 0) {
-    const entrypoint = queue.shift()!;
-    if (visitedEntrypoints.has(entrypoint)) {
-      continue;
+  const depsIndex = {entrypointToDeps: new Map<UrlString, Set<UrlString>>()};
+  const analysis = await analyzer.analyze(entrypoints);
+  const allEntrypoints = new Set<UrlString>(entrypoints);
+  for (const entrypoint of allEntrypoints) {
+    const document = analysis.getDocument(entrypoint);
+    if (!(document instanceof Document)) {
+      const message = document && document.message || '';
+      throw new Error(`Unable to get document ${entrypoint}: ${message}`);
```

Comment:
I don't think we should throw in this case. Ideally we only throw for program errors, and bad user input is handled in normal execution and warnings and errors are threaded through the results. Here we should continue building if possible and just warn and exit with a non-0 code. It would be a bit of work to do that now, probably changing the return type and such, so maybe we can just `console.warn` for now.